### PR TITLE
[QA-Auto] Automate context menu component

### PR DESCRIPTION
**Pull request by WDaniszewska**

### CODE REVIEW

**Review by KlaudiaLapcionek — DISMISSED** [review]
Lgtm :)

File: src/test/java/com/intive/template/pages/ContextMenuPage.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+
+import java.util.List;
+import java.time.Duration;
+
+public class ContextMenuPage extends BasePage {
+    WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(30));
+
+    @FindBy(how = How.CSS, using = "#__next > div > main > div.sc-3c8adf34-0.isvBOn > div:nth-child(2) > div > div > button > span")
```
Comment (MarekPawlowskiQA) [inline src/test/java/com/intive/template/pages/ContextMenuPage.java:16]: Most probably this number will change, whenever new web app is deployed. Try to find different selector. 
But plus for CSS : P

File: src/test/java/com/intive/template/pages/ContextMenuPage.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    WebElement triggerButton;
+    @FindBy(how = How.XPATH, using = "/html/body/div/div/main/div[1]/div[2]/div/div/nav")
+    WebElement openContextMenu;
+    @FindBy(how = How.XPATH, using = "/html/body/div/div/main/div[1]/div[2]/div/div/nav/ul")
+    WebElement listOfItems;
+    @FindBy(how = How.CSS, using = "#__next > div > main > div.sc-3c8adf34-0.isvBOn > div:nth-child(2) > div > div > nav > ul > li:nth-child(1)")
```
Comment (MarekPawlowskiQA) [inline src/test/java/com/intive/template/pages/ContextMenuPage.java:22]: Same here, in NEXT JS framework this values change, and are generated "randomly"

File: src/test/java/com/intive/template/pages/ContextMenuPage.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    public void clickOnItem() {
+        wait.until(ExpectedConditions.visibilityOf(listOfItems));
+        item.click();
+    }
+
+    public void clickOutsideTheContextMenu() {
+        placeOutsideContextMenu.click();
+    }
+
+    //returns true if the context menu is closed and false if it is displayed
```
Comment (MarekPawlowskiQA) [inline src/test/java/com/intive/template/pages/ContextMenuPage.java:57]: unnecessary :) comments are ok, but in this case method is self explanatory

File: src/test/java/com/intive/template/pages/ContextMenuPage.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    public boolean contextMenuShouldBeClose() {
```
Comment (MarekPawlowskiQA) [inline src/test/java/com/intive/template/pages/ContextMenuPage.java:58]: isContectMenuClosed() something is or is not

Comment (MarekPawlowskiQA) [reply src/test/java/com/intive/template/pages/ContextMenuPage.java:58]: btw not many people in IT know this but..... in some way it is related to this problem https://en.wikipedia.org/wiki/Is–ought_problem : D

Comment (MarekPawlowskiQA) [reply src/test/java/com/intive/template/pages/ContextMenuPage.java:58]: But still in method names, pure facts, not wishfull words are used.

File: src/test/java/com/intive/template/pages/ContextMenuPage.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    public void clickOnTriggerButton() {
+        wait.until(ExpectedConditions.visibilityOf(triggerButton));
+        triggerButton.click();
+    }
+
+    public boolean contextMenuIsOpen() {
+        return openContextMenu.isDisplayed();
+    }
+
+    public boolean listContainsTwoItems() {
```
Comment (MarekPawlowskiQA) [inline src/test/java/com/intive/template/pages/ContextMenuPage.java:36]: isListContainsTwoItems()

File: src/test/java/com/intive/template/pages/ContextMenuPage.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    WebElement item;
+    @FindBy(how = How.CLASS_NAME, using = "container")
+    WebElement placeOutsideContextMenu;
+
+    public void clickOnTriggerButton() {
+        wait.until(ExpectedConditions.visibilityOf(triggerButton));
+        triggerButton.click();
+    }
+
+    public boolean contextMenuIsOpen() {
```
Comment (MarekPawlowskiQA) [inline src/test/java/com/intive/template/pages/ContextMenuPage.java:32]: isContectMenuOpen

File: src/test/java/com/intive/template/pages/ContextMenuPage.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        return openContextMenu.isDisplayed();
+    }
+
+    public boolean listContainsTwoItems() {
+        List<WebElement> optionsList = listOfItems.findElements(By.cssSelector("li"));
+        int i = optionsList.size();
```
Comment (MarekPawlowskiQA) [inline src/test/java/com/intive/template/pages/ContextMenuPage.java:38]: I would go with more informative variable

File: src/test/java/com/intive/template/pages/ContextMenuPage.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        System.out.println(i);
+        for (WebElement li : optionsList) {
+            if (!li.isDisplayed() || i != 2) {
```
Comment (MarekPawlowskiQA) [inline src/test/java/com/intive/template/pages/ContextMenuPage.java:41]: Hmmin this case maybe small comment would be needed, for example why this is used i != 2 ?

**Review by MarekPawlowskiQA — CHANGES_REQUESTED** [review]
Requested changes in comments

File: src/test/java/com/intive/template/pages/ContextMenuPage.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+
+import java.util.List;
+import java.time.Duration;
+
+public class ContextMenuPage extends BasePage {
+    WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(30));
```
[the remark this answers is not in the document]
Comment (WDaniszewska) [reply src/test/java/com/intive/template/pages/ContextMenuPage.java]: I changed it as you wrote.

File: src/test/java/com/intive/template/pages/ContextMenuPage.java (hunk 1 above)
Comment (WDaniszewska) [reply src/test/java/com/intive/template/pages/ContextMenuPage.java]: I found another selector.

File: src/test/java/com/intive/template/pages/ContextMenuPage.java (hunk 2 above)
Comment (WDaniszewska) [reply src/test/java/com/intive/template/pages/ContextMenuPage.java]: done.

File: src/test/java/com/intive/template/pages/ContextMenuPage.java (hunk 3 above)
Comment (WDaniszewska) [reply src/test/java/com/intive/template/pages/ContextMenuPage.java]: I removed this comment.

File: src/test/java/com/intive/template/pages/ContextMenuPage.java (hunk 4 above)
Comment (WDaniszewska) [reply src/test/java/com/intive/template/pages/ContextMenuPage.java]: I fix it :)

File: src/test/java/com/intive/template/pages/ContextMenuPage.java (hunk 5 above)
Comment (WDaniszewska) [reply src/test/java/com/intive/template/pages/ContextMenuPage.java]: done

File: src/test/java/com/intive/template/pages/ContextMenuPage.java (hunk 6 above)
Comment (WDaniszewska) [reply src/test/java/com/intive/template/pages/ContextMenuPage.java]: done

File: src/test/java/com/intive/template/pages/ContextMenuPage.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    public boolean contextMenuIsOpen() {
+        return openContextMenu.isDisplayed();
+    }
+
+    public boolean listContainsTwoItems() {
+        List<WebElement> optionsList = listOfItems.findElements(By.cssSelector("li"));
+        int i = optionsList.size();
+        System.out.println(i);
```
[the remark this answers is not in the document]
Comment (WDaniszewska) [reply src/test/java/com/intive/template/pages/ContextMenuPage.java]: done

File: src/test/java/com/intive/template/pages/ContextMenuPage.java (hunk 7 above)
Comment (WDaniszewska) [reply src/test/java/com/intive/template/pages/ContextMenuPage.java]: I changed it

File: src/test/java/com/intive/template/pages/ContextMenuPage.java (hunk 8 above)
Comment (WDaniszewska) [reply src/test/java/com/intive/template/pages/ContextMenuPage.java]: I added a comment

File: src/test/java/com/intive/template/pages/ContextMenuPage.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+        wait.until(ExpectedConditions.visibilityOf(triggerButton));
+        triggerButton.click();
+    }
+
+    public boolean isContextMenuOpen() {
+        return openContextMenu.isDisplayed();
+    }
+
+    //returns true if the context menu contains 2 items or false if it contains less or more than 2 items
+    public boolean isListContainsTwoItems() {
```
Comment (MarekPawlowskiQA) [inline src/test/java/com/intive/template/pages/ContextMenuPage.java:37]: You can make this function more general,  isListContainsItems(itemsNumber) and than, decide how many items you want to check.

Comment (WDaniszewska) [reply src/test/java/com/intive/template/pages/ContextMenuPage.java]: I made changes :)

**Review by MarekPawlowskiQA — APPROVED** [review]
lgtm :)

**Review by KlaudiaLapcionek — APPROVED** [review]
Lgtm :)

**Review by ospatrycja — APPROVED** [review]
lgtm :)

**Pull request merged**